support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -35,22 +37,70 @@ object GiftRecipient {
 
   implicit val e1 = circeDiscriminator.add(deriveEncoder[WeeklyGiftRecipient], discriminatorWeekly)
   implicit val d1 = deriveDecoder[WeeklyGiftRecipient]
-  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubGiftRecipient], discriminatorDigiSub)
-  implicit val d2 = deriveDecoder[DigitalSubGiftRecipient]
+  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubscriptionGiftRecipient], discriminatorDigitalSubscription)
+  implicit val d2 = deriveDecoder[DigitalSubscriptionGiftRecipient]
 
   implicit val encoder: Encoder[GiftRecipient] = Encoder.instance {
     case c: WeeklyGiftRecipient => c.asJson
-    case c: DigitalSubGiftRecipient => c.asJson
+    case c: DigitalSubscriptionGiftRecipient => c.asJson
   }
 
   implicit val decoder: Decoder[GiftRecipient] =
     circeDiscriminator.decode[GiftRecipient](Map(
       discriminatorWeekly -> Decoder[WeeklyGiftRecipient],
-      discriminatorDigiSub -> Decoder[DigitalSubGiftRecipient]
+      discriminatorDigitalSubscription -> Decoder[DigitalSubscriptionGiftRecipient]
+    ))
+}
+
+case class GeneratedGiftCode private(value: String) extends AnyVal
+
+object GeneratedGiftCode {
+
+  def apply(value: String): Option[GeneratedGiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GeneratedGiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GeneratedGiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GeneratedGiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftPurchase {
+  def asDigitalSubscriptionGiftPurchase: Option[DigitalSubscriptionGiftPurchase] =
+    this match { case a: DigitalSubscriptionGiftPurchase => Some(a); case _ => None }
+  def giftRecipient: GiftRecipient
+}
+object GiftPurchase {
+
+  case class WeeklyGiftPurchase(giftRecipient: WeeklyGiftRecipient) extends GiftPurchase
+  case class DigitalSubscriptionGiftPurchase(
+    giftRecipient: DigitalSubscriptionGiftRecipient,
+    giftCode: GeneratedGiftCode,
+  ) extends GiftPurchase
+
+  val circeDiscriminator = new CirceDiscriminator("giftPurchaseType")
+  private val discriminatorWeekly = "Weekly"
+  private val discriminatorDigitalSubscription = "DigitalSubscription"
+
+  implicit val e1 = circeDiscriminator.add(deriveEncoder[WeeklyGiftPurchase], discriminatorWeekly)
+  implicit val d1 = deriveDecoder[WeeklyGiftPurchase]
+  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubscriptionGiftPurchase], discriminatorDigitalSubscription)
+  implicit val d2 = deriveDecoder[DigitalSubscriptionGiftPurchase]
+
+  implicit val encoder: Encoder[GiftPurchase] = Encoder.instance {
+    case c: WeeklyGiftPurchase => c.asJson
+    case c: DigitalSubscriptionGiftPurchase => c.asJson
+  }
+
+  implicit val decoder =
+    circeDiscriminator.decode[GiftPurchase](Map(
+      discriminatorWeekly -> Decoder[WeeklyGiftPurchase],
+      discriminatorDigitalSubscription -> Decoder[DigitalSubscriptionGiftPurchase]
     ))
```

Comment:
I don't really get why it should be so much harder to serialise a sum type than a product type.  I can see why, given how they are implemented in circe, but I feel like there's so much boilerplate.  It almost puts me off modelling things this way.
Should we consider the provided `withDiscriminator` ? https://circe.github.io/circe/codecs/adt.html#the-future